utils/disk_space_monitor.hh

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include <chrono>
+#include <filesystem>
+#include <set>
+
+#include <boost/intrusive/list.hpp>
+
+#include <seastar/core/gate.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/timer.hh>
+#include <seastar/util/optimized_optional.hh>
+
+#include "seastarx.hh"
+
+using namespace std::chrono_literals;
+
+namespace bi = boost::intrusive;
+
+namespace utils {
+
+// Instantiated only on shard 0
+class disk_space_monitor {
+public:
+    struct config {
+        scheduling_group sched_group;
+        lowres_clock::duration normal_polling_interval = 10s;
+        lowres_clock::duration high_polling_interval = 1s;
+        // Use high_polling_interval above this threshold
+        float polling_interval_threshold = 0.95;
+    };
+
+    using subscription_callback_type = noncopyable_function<future<> (const disk_space_monitor&)>;
+    class subscription : public bi::list_base_hook<bi::link_mode<bi::auto_unlink>> {
```

Comment:
What about boost::signals2? Aren't you duplicating it here?

Reply:
Hmm, okay.  Can do.  I wasn't aware of it, but I see we already use it e.g. to get snitch updates